RFC0027.md

```diff
@@ -0,0 +1,190 @@
+
+## Example
+```
+#include <pmix.h>
+#include <pmix_common.h>
+
+typedef struct {
+    pthread_mutex_t mutex;
+    pthread_cond_t cond;
+    volatile bool active;
+    pmix_status_t status;
+} mylock_t;
+
+#define DEBUG_CONSTRUCT_LOCK(l)                     \
+    do {                                            \
+        pthread_mutex_init(&(l)->mutex, NULL);      \
+        pthread_cond_init(&(l)->cond, NULL);        \
+        (l)->active = true;                         \
+        (l)->status = PMIX_SUCCESS;                 \
+    } while(0)
+
+#define DEBUG_DESTRUCT_LOCK(l)              \
+    do {                                    \
+        pthread_mutex_destroy(&(l)->mutex); \
+        pthread_cond_destroy(&(l)->cond);   \
+    } while(0)
+
+#define DEBUG_WAIT_THREAD(lck)                                      \
+    do {                                                            \
+        pthread_mutex_lock(&(lck)->mutex);                          \
+        while ((lck)->active) {                                     \
+            pthread_cond_wait(&(lck)->cond, &(lck)->mutex);         \
+        }                                                           \
+        pthread_mutex_unlock(&(lck)->mutex);                        \
+    } while(0)
+
+#define DEBUG_WAKEUP_THREAD(lck)                        \
+    do {                                                \
+        pthread_mutex_lock(&(lck)->mutex);              \
+        (lck)->active = false;                          \
+        pthread_cond_broadcast(&(lck)->cond);           \
+        pthread_mutex_unlock(&(lck)->mutex);            \
+    } while(0)
+
+static void cbfunc(pmix_status_t status,
+                   pmix_info_t *info, size_t ninfo,
+                   void *cbdata,
+                   pmix_release_cbfunc_t release_fn,
+                   void *release_cbdata)
+{
+    mylock_t *lock = (mylock_t*)cbdata;
+
+    lock->status = status;
+
+    /* let the library release the data and cleanup from
+     * the operation */
+    if (NULL != release_fn) {
+        release_fn(release_cbdata);
+    }
+
+    /* release the block */
+    DEBUG_WAKEUP_THREAD(lock);
+}
+
+int main (int argc, char **argv)
+{
+    pmix_proc_t myproc, proc;
+    pmix_info_t *info;
+    mylock_t lock;
+
+    PMIx_Init(&myproc, NULL, 0);
+
+    /* request to cleanup a process-specific tmp directory tree and a specific shmem backing file */
+    PMIX_INFO_CREATE(info, 5);
+    PMIX_INFO_LOAD(&info[0], PMIX_REGISTER_CLEANUP, "/mytmpdir", PMIX_STRING);
+    PMIX_INFO_LOAD(&info[1], PMIX_REGISTER_CLEANUP, "/tmp/dev_shm/mybackfile", PMIX_STRING);
+    /* recursively cleanup subdirectories */
+    PMIX_INFO_LOAD(&info[2], PMIX_CLEANUP_RECURSIVE, NULL, PMIX_BOOL);
```

Comment:
@rhc54 to which directory this recursive flag will be applied?

Reply:
All of the ones contained in this particular registration - if you want one to be recursive and the other not, then they need to be registered separately.